### ISSUE TITLE
roleAssignments returns `HTTP Error 400. The request is badly formed.`

### ISSUE THREAD

**Issue by fals** [Azure/azure-sdk-for-go#20847]
### Bug Report

- github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/authorization/armauthorization/v2
- github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/authorization/armauthorization/v2 v2.1.1
-go version go1.20.4 darwin/arm64

- What happened?
When calling the List for the new API with the same filters as specified in the documentation:
```go
GET https://management.usgovcloudapi.net/subscriptions/{sub_id}/providers/Microsoft.Authorization/roleAssignments
--------------------------------------------------------------------------------
RESPONSE 400: 400 Bad Request
ERROR CODE UNAVAILABLE
--------------------------------------------------------------------------------
<!DOCTYPE HTML PUBLIC \"-//W3C//DTD HTML 4.01//EN\"\"http://www.w3.org/TR/html4/strict.dtd\">\r\n<HTML><HEAD><TITLE>Bad Request</TITLE>\r\n<META HTTP-EQUIV=\"Content-Type\" Content=\"text/html; charset=us-ascii\"></HEAD>\r\n<BODY><h2>Bad Request</h2>\r\n<hr><p>HTTP Error 400. The request is badly formed.</p>\r\n</BODY></HTML>\r\n
--------------------------------------------------------------------------------
```
- What did you expect or want to happen?
Should return the Roles Assignments for the principalId in the filter.
- How can we reproduce it?
```go
opts := azcore.ClientOptions{}
if o.IsAzureGovCloud {
	opts.Cloud = azcloud.AzureGovernment
}

armOpts := arm.ClientOptions{
	ClientOptions: opts,
}
rsClient, err := armauthorization.NewRoleAssignmentsClient(creds.SubscriptionID, creds, &armOpts)
if err != nil {
	return nil, fmt.Errorf("initializing new role assignments client: %w", err)
}
var res []armauthorization.RoleAssignment
f := fmt.Sprintf("principalId eq '%v'", principalID)
pager := rsClient.NewListForSubscriptionPager(
       &armauthorization.RoleAssignmentsClientListForSubscriptionOptions {
		Filter:   &f,
		TenantID: nil,
	})
for pager.More() {
	nextResult, err := pager.NextPage(ctx)
	if err != nil {
		return nil, err
	}
	for _, ra := range nextResult.Value {
		res = append(res, *ra)
	}
}
```

**Comment by jhendrixMSFT** [comment]
Thanks this is helpful.

Per the OpenAPI definition, the [`RoleAssignmentFilterParameter`](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/authorization/resource-manager/Microsoft.Authorization/stable/2022-04-01/authorization-RoleAssignmentsCalls.json#L679-L687) isn't supposed to be URL-encoded.  However, if I fix up the SDK code to _encode_ the filter parameter then it works as expected.

**Comment by mjudeikis** [comment]
@jhendrixMSFT any ETA on this, or do you have a hack we can apply?

**Comment by fals** [comment]
@jhendrixMSFT we have another issue, the [NewListForResourcePager](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/resourcemanager/authorization/armauthorization/roleassignments_client.go#LL363C38-L363C61) it doesn't work in some cases.

```html
-------------------------------------------------------------------------------- 
RESPONSE 400: 400 Bad Request 
ERROR CODE UNAVAILABLE 
-------------------------------------------------------------------------------- 
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN""http://www.w3.org/TR/html4/strict.dtd"> <HTML><HEAD><TITLE>Bad Request</TITLE> <META HTTP-EQUIV="Content-Type" Content="text/html; charset=us-ascii"></HEAD> <BODY><h2>Bad Request</h2> <hr><p>HTTP Error 400. The request is badly formed.</p> </BODY></HTML> 
--------------------------------------------------------------------------------
```

**Comment by jhendrixMSFT** [comment]
I think this should get you unblocked.
```go
f := url.QueryEscape(fmt.Sprintf("principalId eq '%v'", principalID))
```

**Comment by fals** [comment]
@jhendrixMSFT any news on this? The solution you suggested doesn't work, if you do like that the call will work, but the filter won't be applied correctly, the response will contain records that are not from the principal. The code snippet you provided replaces white spaces by `+`, the correct comparing with the REST call I gave as example should be UTF-8 encoded `%20`.

We worked around from our side using rest calls, but quite unfortunate having such version of the sdk going to release like this. 

Is this code all generated from OpenApi spec or can I submit a PR to fix this package?

**Comment by mattwelke** [comment]
Ian into this today and I found a solution. I'm not sure if the core issue is related to the REST endpoint, the Go SDK, or that everything is working fine and it's just a lack of documentation right now. [The official documentation](https://learn.microsoft.com/en-us/rest/api/authorization/role-assignments/list-for-subscription?view=rest-authorization-2022-04-01&tabs=HTTP) lacks an example that uses a `principalId` filter, so there's no way to tell what syntax to use. See details below.

#### Not working reproduction

My code was:

```go
func main() {
	subscriptionID := "<subscription-id>"
	principalID := "<principal-id>"

	cred, _ := azidentity.NewDefaultAzureCredential(nil)
	clientFactory, _ := armauthorization.NewClientFactory(subscriptionID, cred, nil)
	client := clientFactory.NewRoleAssignmentsClient()

	filter := fmt.Sprintf("principalId eq %s", principalID)
	opts := &armauthorization.RoleAssignmentsClientListForSubscriptionOptions{
		Filter: &filter,
	}
	pager := client.NewListForSubscriptionPager(opts)

	var roleAssignments []*armauthorization.RoleAssignment
	for pager.More() {
		nextResult, err := pager.NextPage(context.TODO())
	...
}
```

And this resulted in an error being returned from `pager.NextPage` with a `400` status code and an `"UnsupportedQuery"` error code.

#### Solution

When trying different workarounds, including the ones suggested here, I found that these two techniques made it work:

* URL-encode the filter string with `url.QueryEscape`, *and* surround the principal ID value with `'`.
* URL-encode the filter string manually by replacing spaces in the `fmt.Sprintf` argument with the substring `%%20` (to escape the `%`), *and* surround the principal ID value with `'`.

So, I think the missing piece here was to surround the principal ID with the single quotes.

This code ended up working for me:

```go
func main() {
	subscriptionID := "<subscription-id>"
	principalID := "<principal-id>"

	cred, _ := azidentity.NewDefaultAzureCredential(nil)
	clientFactory, _ := armauthorization.NewClientFactory(subscriptionID, cred, nil)
	client := clientFactory.NewRoleAssignmentsClient()

	filter := url.QueryEscape(fmt.Sprintf("principalId eq '%s'", principalID))
	opts := &armauthorization.RoleAssignmentsClientListForSubscriptionOptions{
		Filter: &filter,
	}
	pager := client.NewListForSubscriptionPager(opts)

	var roleAssignments []*armauthorization.RoleAssignment
	for pager.More() {
		nextResult, err := pager.NextPage(context.TODO())
	...
}
```

(I preferred using `url.QueryEscape` over manually using `%%20`)

https://github.com/Azure/azure-sdk-for-go/issues/20847#issuecomment-1568017764 mentioned that their query wasn't working correctly because even though the request was successful, it returned more role assignments than it should have, thus not filtering at all. In my code that worked, my filter *did* restrict the role assignments correctly. It did not return more than it should have.

**Comment by jhendrixMSFT** [comment]
Thanks for sharing your solution, this is helpful. I've seen other examples (can't find it right now) where the filtered expression is enclosed in single quotes, so I suspect that this is the expected format.

@sandeep-sen do you know who we can follow up with on the service team about the expected format? In addition, the docs should be updated with an example.

**Comment by mattwelke** [comment]
Good timing. I was just filling out a documentation feedback form. I decided to just send them here for full context now that it's been acknowledged.

![image](https://github.com/Azure/azure-sdk-for-go/assets/7719209/a4c1d20d-23eb-4dc1-9434-6fddc748ce06)

**Comment by mattwelke** [comment]
Adding a note here that this issue does not affect the [deny assignments equivalent](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/authorization/armauthorization/v2@v2.2.0#DenyAssignmentsClient.NewListForScopePager). Therefore, if you try to use the same string (where you manually escape characters) for [Filter](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/authorization/armauthorization/v2@v2.2.0#DenyAssignmentsClientListForScopeOptions.Filter), it will return a 400. As of right now, you must not escape the string in order to get deny assignments and you must escape the string in order to get role assignments.

Examples:

```go
// Get the deny assignments at the scope for the principal.
filter := fmt.Sprintf("principalId eq '%s'", principalID)
daPager := daClient.NewListForScopePager(scope, &armauthorization.DenyAssignmentsClientListForScopeOptions{
	Filter: &filter,
})
```

```go
// Get the role assignments at the scope for the principal.
filter := url.QueryEscape(fmt.Sprintf("principalId eq '%s'", principalID))
raPager := raClient.NewListForScopePager(scope, &armauthorization.RoleAssignmentsClientListForScopeOptions{
	Filter: &filter,
})
```
